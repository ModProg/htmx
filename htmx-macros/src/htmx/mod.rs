mod html;
mod special_components;

mod rusty;

pub use html::html;
pub use rusty::html as rtml;

// pub fn html(input: TokenStream) -> Result {
//     if input.is_empty() {
//         return Ok(quote!(::htmx::Html::new()));
//     }

//     let mut fork = input.clone().into_iter();

//     let first = fork.next();
//     let second = fork.next();

//     // TODO figure out actual differentiator
//     // probably would be, starts with `<` or starts with `{}` or `""` not
// followed     // by `,`

//     if matches!(input.peek(), Some(TokenTree::Punct(punct)) if
// punct.as_char() == '<') {         html::html(input.collect())
//     } else {
//         rusty::html(input.collect())
//     }
// }
