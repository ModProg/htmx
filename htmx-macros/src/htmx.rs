use htmx_script::{Script, ToJs};
use manyhow::{ensure, Result};
use proc_macro2::{TokenStream, TokenTree};
use quote::ToTokens;
use rstml::atoms::{CloseTag, OpenTag};
use rstml::node::{
    AttributeValueExpr, KeyedAttribute, KeyedAttributeValue, NodeAttribute, NodeBlock, NodeElement,
    NodeName,
};
use rstml::recoverable::Recoverable;
use syn::spanned::Spanned;
use syn::{parse2, Expr, ExprLit, ExprPath, Lit, LitStr, Stmt};

use crate::special_components::Node;
use crate::*;

pub fn htmx(input: TokenStream) -> Result {
    // https://github.com/rust-lang/rust-analyzer/issues/15572
    // let htmx = match (
    //     proc_macro_crate::crate_name("htmx"),
    //     std::env::var("CARGO_CRATE_NAME").as_deref(),
    // ) { (Ok(FoundCrate::Itself), Ok("htmx")) => quote!(crate),
    //   (Ok(FoundCrate::Name(name)), _) => { let ident = Ident::new(&name,
    //   Span::call_site()); quote!(::#ident) } _ => quote!(::htmx),
    // };
    let htmx = htmx_crate();

    let mut input = input.into_iter().peekable();

    let htmx = match input.peek() {
        Some(TokenTree::Ident(ident)) if ident == "crate" => {
            input.next();
            quote!(crate)
        }
        _ => htmx,
    };

    if input.peek().is_none() {
        return Ok(quote!(#htmx::Html::new()));
    }

    let nodes = rstml::Parser::new(
        rstml::ParserConfig::new()
            .recover_block(true)
            .element_close_use_default_wildcard_ident(false)
            .raw_text_elements(["script"].into())
            .custom_node(),
    )
    // TODO parse_recoverable
    .parse_simple(input.collect::<TokenStream>())?
    .into_iter()
    .map(|n| expand_node(n, &htmx, false))
    .collect::<Result>()?;

    Ok(quote! {
        #use #htmx::{ToHtml, Html, IntoHtmlElements};
        {
            use #htmx::native::*;
            let mut $htmx = Html::new();
            #nodes
            $htmx
        }
    })
}

pub fn expand_node(node: Node, htmx: &TokenStream, child: bool) -> Result {
    Ok(match node {
        Node::Comment(_) => todo!("{}", line!()),
        Node::Doctype(_) => todo!("{}", line!()),
        Node::Fragment(_) => todo!("{}", line!()),
        Node::Element(NodeElement {
            open_tag: OpenTag {
                name, attributes, ..
            },
            children,
            close_tag,
            ..
        }) => {
            let script = name.to_string() == "script";
            let (name, custom) = name_to_struct(name, htmx)?;
            let attributes = attributes
                .into_iter()
                .map(|attribute| match attribute {
                    NodeAttribute::Block(attr) => Ok(quote!(custom_attr(#attr, true))),
                    NodeAttribute::Attribute(KeyedAttribute {
                        key,
                        possible_value,
                    }) => match possible_value {
                        KeyedAttributeValue::Binding(_) => todo!("{}", line!()),
                        KeyedAttributeValue::Value(AttributeValueExpr { value, .. }) => {
                            attribute_key_to_fn(key, value, custom)
                        }
                        KeyedAttributeValue::None => attribute_key_to_fn(key, true, custom),
                    },
                })
                .collect::<Result<Vec<_>>>()?;
            let children = if children.is_empty() {
                quote!()
            } else if script {
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    quote!($node = $node.child(#script);)
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    quote!($node = $node.child({#[allow(unused_braces)] #block});)
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    quote!($node = $node.child(#script);)
                }
            } else {
                expand_nodes(children, htmx, true)?
            };
            let main = quote!({let mut $node = #name #(.#attributes)*; #children $node.build()});

            let main = match close_tag {
                Some(CloseTag {
                    name: name @ NodeName::Path(_),
                    ..
                }) if !name.is_wildcard() => {
                    // If close_tag was specified, use it so coloring happens
                    quote!({let _: #name; #main})
                }
                _ => main,
            };
            if child {
                quote!($node = $node.child(&#main);)
            } else {
                quote!(#htmx::ToHtml::write_to_html(&#main, &mut $htmx);)
            }
        }
        Node::Block(_) | Node::Text(_) => {
            if child {
                quote!($node = $node.child(&{#[allow(unused_braces)] #node});)
            } else {
                quote!(#htmx::ToHtml::write_to_html(&{#[allow(unused_braces)] #node}, &mut $htmx);)
            }
        }
        Node::RawText(_) => todo!("{}", line!()),
        Node::Custom(c) => c.expand_node(htmx, child)?,
    })
}

pub fn expand_nodes(nodes: Vec<Node>, htmx: &TokenStream, child: bool) -> Result {
    nodes
        .into_iter()
        .map(|n| expand_node(n, htmx, child))
        .collect()
}

fn name_to_struct(name: NodeName, htmx: &TokenStream) -> Result<(TokenStream, bool)> {
    match name {
        NodeName::Path(path) => Ok((quote!(#path::builder()), false)),
        name @ NodeName::Punctuated(_) => {
            let name = name.to_string();
            ensure!(name.to_ascii_lowercase().chars().all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
         "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
        );
            Ok((quote!(#htmx::CustomElement::new_unchecked(#name)), true))
        }
        // This {...}
        NodeName::Block(name) => {
            if let [
                Stmt::Expr(
                    Expr::Lit(ExprLit {
                        lit: Lit::Str(name),
                        ..
                    }),
                    None,
                ),
            ] = &name.stmts[..]
            {
                let name = name.value();
                ensure!(name.to_ascii_lowercase().chars().all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
                 "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
                );
                Ok((quote!(#htmx::CustomElement::new_unchecked(#name)), true))
            } else {
                Ok((quote!(#htmx::CustomElement::new(#name)), true))
            }
        }
    }
}

fn attribute_key_to_fn(name: NodeName, value: impl ToTokens, custom: bool) -> Result {
    Ok(match name {
        NodeName::Path(ExprPath { path, .. }) if !custom && path.get_ident().is_some() => {
            quote!(#path(#value))
        }
        NodeName::Path(ExprPath { path, .. })
            if path.segments.first().is_some_and(|hx| hx.ident == "hx") =>
        {
            {
                let sident = path
                    .segments
                    .iter()
                    .map(|i| i.ident.to_string().replace('_', "-"))
                    // hx::swap::oob
                    .collect::<Vec<_>>()
                    .join("-");
                quote_spanned!(path.span()=> custom_attr(#sident, #value))
            }
        }
        // This {...}
        name @ (NodeName::Punctuated(_) | NodeName::Path(_)) => {
            let sname = name.to_string();
            quote_spanned!(name.span()=>  custom_attr(#sname, #value))
        }
        name @ NodeName::Block(_) => quote_spanned!(name.span()=>  custom_attr(#name, #value)),
    })
}
