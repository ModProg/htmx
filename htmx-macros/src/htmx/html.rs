use htmx_script::{Script, ToJs};
use manyhow::{ensure, ErrorMessage, Result};
use proc_macro2::TokenStream;
use proc_macro_utils::TokenStream2Ext;
use quote::ToTokens;
use rstml::atoms::{CloseTag, OpenTag};
use rstml::node::{
    AttributeValueExpr, KeyedAttribute, KeyedAttributeValue, NodeAttribute, NodeBlock, NodeElement,
    NodeName,
};
use rstml::recoverable::Recoverable;
use syn::spanned::Spanned;
use syn::{parse2, Expr, ExprLit, ExprPath, Lit, LitStr, Stmt};

use super::special_components::{Node, Special};
use crate::*;

pub fn html(input: TokenStream) -> Result {
    let mut parser = input.clone().parser();
    let (input, expr) =
        if let ( Some(expr), Some(_)) = (parser.next_expression(), parser.next_tt_fat_arrow()) {
            (parser.into_token_stream(), quote!(&mut #expr))
        } else {
            (input, quote!(::htmx::Html::new()))
        };

    let nodes = rstml::Parser::new(
        rstml::ParserConfig::new()
            .recover_block(true)
            .element_close_use_default_wildcard_ident(false)
            .custom_node::<Special>()
            .raw_text_elements(["script"].into()),
    )
    // TODO parse_recoverable
    .parse_simple(input)?
    .into_iter()
    .map(expand_node)
    .collect::<Result>()?;

    Ok(quote! {
        {
            use ::htmx::native::*;
            let mut __html = #expr;
            #nodes
            __html
        }
    })
}

pub fn expand_node(node: Node) -> Result {
    Ok(match node {
        Node::Comment(_) => todo!("{}", line!()),
        Node::Doctype(_) => todo!("{}", line!()),
        Node::Fragment(_) => todo!("{}", line!()),
        Node::Element(NodeElement {
            open_tag: OpenTag {
                name, attributes, ..
            },
            children,
            close_tag,
            ..
        }) => {
            let script = name.to_string() == "script";
            let (name, custom) = name_to_struct(name)?;
            let attributes = attributes
                .into_iter()
                .map(|attribute| match attribute {
                    NodeAttribute::Block(attr) => Ok(quote!(custom_attr(#attr, true))),
                    NodeAttribute::Attribute(KeyedAttribute {
                        key,
                        possible_value,
                    }) => match possible_value {
                        KeyedAttributeValue::Binding(_) => todo!("{}", line!()),
                        KeyedAttributeValue::Value(AttributeValueExpr { value, .. }) => {
                            attribute_key_to_fn(key, value, custom)
                        }
                        KeyedAttributeValue::None => attribute_key_to_fn(key, true, custom),
                    },
                })
                .collect::<Result<Vec<_>>>()?;
            let children = if children.is_empty() {
                quote!()
            } else if script {
                // TODO scripts
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    quote!(__html.child_expr(#script);)
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    quote!(__html.child_expr({#[allow(unused_braces)] #block});)
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    quote!(__html.child(#script);)
                }
            } else {
                expand_nodes(children)?
            };
            let body = (!children.is_empty()).then(|| quote!(let mut __html = __html.body();));
            let main = quote!({let mut __html = #name #(__html.#attributes;)* #body; #children});

            match close_tag {
                Some(CloseTag {
                    name: name @ NodeName::Path(_),
                    ..
                }) if !name.is_wildcard() => {
                    // If close_tag was specified, use it so coloring happens
                    quote!({#name::unused(); #main})
                }
                _ => main,
            }
        }
        Node::Block(_) | Node::Text(_) => {
            quote!(::htmx::ToHtml::to_html(&{#[allow(unused_braces)] #node}, &mut __html);)
        }
        Node::RawText(_) => todo!("{}", line!()),
        Node::Custom(c) => c.expand_node()?,
    })
}

pub fn expand_nodes(nodes: Vec<Node>) -> Result {
    nodes.into_iter().map(expand_node).collect()
}

pub fn ensure_tag_name(name: String, span: impl ToTokens) -> Result<String, ErrorMessage> {
    ensure!(
        name.to_ascii_lowercase().chars()
            .all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
        span,
        "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
        // TODO similar function but with css error: https://drafts.csswg.org/css-syntax-3/#non-ascii-ident-code-point
    );
    Ok(name)
}

fn name_to_struct(name: NodeName) -> Result<(TokenStream, bool)> {
    match name {
        NodeName::Path(path) => Ok((quote!(#path::new(&mut __html);), false)),
        name @ NodeName::Punctuated(_) => {
            let name = ensure_tag_name(name.to_string(), name)?;
            Ok((
                quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name);),
                true,
            ))
        }
        // This {...}
        NodeName::Block(name) => {
            if let [
                Stmt::Expr(
                    Expr::Lit(ExprLit {
                        lit: Lit::Str(name),
                        ..
                    }),
                    None,
                ),
            ] = &name.stmts[..]
            {
                let name = ensure_tag_name(name.value(), name)?;
                Ok((
                    quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name);),
                    true,
                ))
            } else {
                Ok((
                    quote!(::htmx::CustomElement::new(&mut __html, #name);),
                    true,
                ))
            }
        }
    }
}

fn attribute_key_to_fn(name: NodeName, value: impl ToTokens, custom: bool) -> Result {
    Ok(match name {
        NodeName::Path(ExprPath { path, .. }) if !custom && path.get_ident().is_some() => {
            quote!(#path(#value))
        }
        NodeName::Path(ExprPath { path, .. })
            if path.segments.first().is_some_and(|hx| hx.ident == "hx") =>
        {
            {
                let sident = path
                    .segments
                    .iter()
                    .map(|i| i.ident.to_string().replace('_', "-"))
                    // hx::swap::oob
                    .collect::<Vec<_>>()
                    .join("-");
                quote_spanned!(path.span()=> custom_attr(#sident, #value))
            }
        }
        // This {...}
        name @ (NodeName::Punctuated(_) | NodeName::Path(_)) => {
            let sname = name.to_string();
            quote_spanned!(name.span()=>  custom_attr(#sname, #value))
        }
        name @ NodeName::Block(_) => quote_spanned!(name.span()=>  custom_attr(#name, #value)),
    })
}
