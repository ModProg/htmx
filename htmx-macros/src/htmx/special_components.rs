use std::borrow::Borrow;
use std::iter;

use manyhow::Result;
use proc_macro2::TokenStream;
use quote::{ToTokens, TokenStreamExt};
use rstml::node::CustomNode;
use rstml::recoverable::{ParseRecoverable, RecoverableContext};
use syn::parse::{ParseBuffer, ParseStream};
use syn::punctuated::Punctuated;
use syn::token::{Brace, Paren};
use syn::{Expr, ExprPath, Token};
use syn_derive::ToTokens;

use super::html::{expand_node, expand_nodes};
use crate::*;
pub type Node = rstml::node::Node<Special>;

macro_rules! braced {($name:ident in $parser:expr, $input:expr) => {{
    let brace = $parser.save_diagnostics((|| {
        let content;
        let brace = syn::braced!(content in $input);
        Ok((brace, content))
        })())?;
    $name = brace.1;
    brace.0
}};}

macro_rules! parenthesized {($name:ident in $parser:expr, $input:expr) => {{
    let paren = $parser.save_diagnostics((|| {
        let content;
        let paren = syn::parenthesized!(content in $input);
        Ok((paren, content))
        })())?;
    $name = paren.1;
    paren.0
}};}

fn parse_nodes<'a>(
    parser: &mut RecoverableContext,
    input: impl Borrow<ParseBuffer<'a>>,
) -> Vec<Node> {
    iter::from_fn(|| parser.parse_recoverable(input.borrow())).collect()
}

///// Unsure how to end the `if`, e.g., in the case of and `else` / `else if`
// <if a> ... </if>
// <if let Some(a) = a> ... </if>
// <if a> ... <else> ... </if>
// <if a> ... <else if b> ... </if>

// <for a in b> ... </for>
// <while a> ... </while>
// <while let Some(a) = b.next()> ... </while>

///// Unsure what syntax to use for match arms
// <match a>
//     <Some(b)> ... </>
//     <None> ... </>
//     <_> ... </>
///// OR
//     <case Some(b)> ... </case>
//     <case None> ... </case>
//     <default> ... </default>
// </match>

// TODO consider using non tag control flow

#[derive(Debug, ToTokens)]
pub enum Special {
    If(If),
    For(For),
    While(While),
    FunctionCall(FunctionCall),
}

fn map_vec(value: Vec<Node>) -> Result<Vec<super::Node>> {
    value.into_iter().map(super::Node::try_from).collect()
}

impl TryFrom<Special> for super::Node {
    type Error = manyhow::Error;

    fn try_from(value: Special) -> std::result::Result<Self, Self::Error> {
        Ok(match value {
            Special::If(if_) => super::Node::If(if_.try_into()?),
            Special::For(For {
                pat, expr, body, ..
            }) => super::Node::For(super::For {
                pat: pat.into_token_stream(),
                expr: expr.into_token_stream(),
                body: map_vec(body)?,
            }),
            Special::While(While { expr, body, .. }) => super::Node::While(super::While {
                expr: expr.into_token_stream(),
                body: map_vec(body)?,
            }),
            Special::FunctionCall(FunctionCall { function, args, .. }) => {
                super::Node::FunctionCall(super::FunctionCall {
                    function: function.into_token_stream(),
                    args: args.into_iter().map(ToTokens::into_token_stream).collect(),
                })
            }
        })
    }
}

impl Special {
    pub(crate) fn expand_node(self) -> Result {
        match self {
            Special::If(if_) => if_.expand_node(),
            Special::For(for_) => for_.expand_node(),
            Special::While(while_) => while_.expand_node(),
            Special::FunctionCall(function_call) => function_call.expand_node(),
        }
    }
}

impl CustomNode for Special {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        ToTokens::to_tokens(self, tokens)
    }

    fn peek_element(input: ParseStream) -> bool {
        let fork = input.fork();
        input.peek(Token![if])
            || input.peek(Token![for])
            || input.peek(Token![while])
            || fork.parse::<Token![<]>().is_ok()
                && fork.parse::<ExprPath>().is_ok()
                && fork.peek(Paren)
    }

    fn parse_element(parser: &mut RecoverableContext, input: ParseStream) -> Option<Self> {
        match () {
            () if input.peek(Token![if]) => parser.parse_recoverable(input).map(Self::If),
            () if input.peek(Token![for]) => parser.parse_recoverable(input).map(Self::For),
            () if input.peek(Token![while]) => parser.parse_recoverable(input).map(Self::While),
            () if input.peek(Token![<]) => parser.parse_recoverable(input).map(Self::FunctionCall),
            _ => unreachable!("`peek_element` should only peek valid keywords"),
        }
    }
}

#[derive(Debug, ToTokens)]
pub struct If {
    pub if_token: Token![if],
    pub condition: Expr,
    #[syn(braced)]
    pub brace: Brace,
    #[syn(in = brace)]
    #[to_tokens(TokenStreamExt::append_all)]
    pub then_branch: Vec<Node>,
    pub else_branch: ElseBranch,
}

impl TryFrom<If> for super::If {
    type Error = manyhow::Error;

    fn try_from(
        If {
            condition,
            then_branch,
            else_branch,
            ..
        }: If,
    ) -> std::result::Result<Self, Self::Error> {
        Ok(super::If {
            condition: condition.into_token_stream(),
            then_branch: map_vec(then_branch)?,
            else_branch: match else_branch {
                ElseBranch::None => super::ElseBranch::None,
                ElseBranch::Else { body, .. } => super::ElseBranch::Else(map_vec(body)?),
                ElseBranch::ElseIf { body, .. } => {
                    super::ElseBranch::ElseIf(Box::new((*body).try_into()?))
                }
            },
        })
    }
}

impl If {
    fn expand_node(self) -> Result {
        let If {
            if_token,
            condition,
            then_branch,
            else_branch,
            ..
        } = self;
        let body = then_branch
            .into_iter()
            .map(expand_node)
            .collect::<Result>()?;
        let else_branch = else_branch.expand_node()?;
        Ok(quote! {
            #if_token #condition {
                #body
            } #else_branch
        })
    }
}

#[derive(Debug, ToTokens)]
pub enum ElseBranch {
    None,
    Else {
        else_token: Token![else],
        #[syn(braced)]
        brace: Brace,
        #[syn(in = brace)]
        #[to_tokens(TokenStreamExt::append_all)]
        body: Vec<Node>,
    },
    ElseIf {
        else_token: Token![else],
        body: Box<If>,
    },
}
impl ElseBranch {
    fn expand_node(self) -> Result {
        Ok(match self {
            ElseBranch::None => quote!(),
            ElseBranch::Else {
                else_token, body, ..
            } => {
                let body = expand_nodes(body)?;
                quote!( #else_token {#body} )
            }
            ElseBranch::ElseIf { else_token, body } => {
                let body = body.expand_node()?;
                quote!(#else_token #body)
            }
        })
    }
}

impl ParseRecoverable for If {
    fn parse_recoverable(parser: &mut RecoverableContext, input: ParseStream) -> Option<Self> {
        let body;
        Some(Self {
            if_token: parser.parse_simple(input)?,
            condition: parser.save_diagnostics(Expr::parse_without_eager_brace(input))?,
            brace: braced!(body in parser, input),
            then_branch: parse_nodes(parser, body),
            else_branch: if let Ok(else_token) = input.parse() {
                if input.peek(Token![if]) {
                    ElseBranch::ElseIf {
                        else_token,
                        body: Box::new(parser.parse_recoverable(input)?),
                    }
                } else {
                    let body;
                    ElseBranch::Else {
                        else_token,
                        brace: braced!(body in parser, input),
                        body: parse_nodes(parser, body),
                    }
                }
            } else {
                ElseBranch::None
            },
        })
    }
}

#[derive(Debug, ToTokens)]
pub struct For {
    pub for_token: Token![for],
    pub pat: syn::Pat,
    pub in_token: Token![in],
    pub expr: Expr,
    #[syn(braced)]
    pub brace: Brace,
    #[syn(in = brace)]
    #[to_tokens(TokenStreamExt::append_all)]
    pub body: Vec<Node>,
}
impl For {
    fn expand_node(self) -> Result {
        let Self {
            for_token,
            pat,
            in_token,
            expr,
            body,
            ..
        } = self;
        let body = expand_nodes(body)?;
        Ok(quote!(#for_token #pat #in_token #expr { #body }))
    }
}

impl ParseRecoverable for For {
    fn parse_recoverable(parser: &mut RecoverableContext, input: ParseStream) -> Option<Self> {
        let body;
        Some(Self {
            for_token: parser.parse_simple(input)?,
            pat: parser.save_diagnostics(syn::Pat::parse_multi_with_leading_vert(input))?,
            in_token: parser.parse_simple(input)?,
            expr: parser.save_diagnostics(Expr::parse_without_eager_brace(input))?,
            brace: braced!(body in parser, input),
            body: parse_nodes(parser, body),
        })
    }
}

#[derive(Debug, ToTokens)]
pub struct While {
    pub while_token: Token![while],
    pub expr: Expr,
    #[syn(braced)]
    pub brace: Brace,
    #[syn(in = brace)]
    #[to_tokens(TokenStreamExt::append_all)]
    pub body: Vec<Node>,
}

impl While {
    fn expand_node(self) -> Result {
        let Self {
            while_token,
            expr,
            body,
            ..
        } = self;
        let body = expand_nodes(body)?;
        Ok(quote!(#while_token #expr { #body }))
    }
}

impl ParseRecoverable for While {
    fn parse_recoverable(parser: &mut RecoverableContext, input: ParseStream) -> Option<Self> {
        let body;
        Some(Self {
            while_token: parser.parse_simple(input)?,
            expr: parser.save_diagnostics(Expr::parse_without_eager_brace(input))?,
            brace: braced!(body in parser, input),
            body: parse_nodes(parser, body),
        })
    }
}

#[derive(Debug, ToTokens)]
pub struct FunctionCall {
    pub open_token: Token![<],
    pub function: ExprPath,
    #[syn(parenthesized)]
    pub paren: Paren,
    #[syn(in = paren)]
    #[to_tokens(TokenStreamExt::append_all)]
    pub args: Punctuated<Expr, Token![,]>,
    pub slash: Token![/],
    pub gt_token: Token![>],
}

impl FunctionCall {
    fn expand_node(self) -> Result {
        let Self { function, args, .. } = self;
        let args = args.into_iter();
        Ok(quote!(::htmx::ToHtml::to_html(&#function(#(Into::into(#args),)*), &mut __html);))
    }
}

impl ParseRecoverable for FunctionCall {
    fn parse_recoverable(parser: &mut RecoverableContext, input: ParseStream) -> Option<Self> {
        let args;
        Some(Self {
            open_token: parser.parse_simple(input)?,
            function: parser.parse_simple(input)?,
            paren: parenthesized!(args in parser, input),
            args: parser.save_diagnostics(Punctuated::parse_terminated(&args))?,
            slash: parser.parse_simple(input)?,
            gt_token: parser.parse_simple(input)?,
        })
    }
}
