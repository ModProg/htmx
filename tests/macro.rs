use htmx::htmx;
use htmx::{Html, ToHtml};

struct Custom {
    href: String,
}

impl Custom {
    fn builder() -> Self {
        Custom {
            href: "default".to_owned(),
        }
    }

    fn href(mut self, value: impl Into<String>) -> Self {
        self.href = value.into();
        self
    }

    fn build(self) -> Html {
        htmx! {
            <a href=self.href/>
        }
    }
}

#[test]
fn test() {
    let html = htmx!(
        <div>
            <a href="hello" download/>
            <a href="hello" download="file.name"/>
            <Custom href="test"/>
        </div>
    );
    insta::assert_snapshot!(html.into_html().to_string());
}
