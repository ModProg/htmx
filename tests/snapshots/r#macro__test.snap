---
source: tests/macro.rs
expression: html.into_html().to_string()
---
<div><a download href="hello"></a><a download="file.name" href="hello"></a><a href="test"></a></div>
